[NEW] Add Feedback form

murtaza98 opened the pull request:
The main goal of this PR is to add functionality to display feedback form to visitors at the end of conversation.
Tasks

- [x] Add a Feedback form
- [ ] connect Feedback form with Router
- [ ] connect feedback form with API

Feedback form UI
![image](https://user-images.githubusercontent.com/34130764/82758388-ea3a4800-9e03-11ea-8157-85b635ce952e.png)


PS:
Most of the code here is taken from PR #104. There have been many changes in this repo since this PR #104 was opened, so I decided to create a new PR instead of updating and merging the old one.

murtaza98 commented:
Hello, @renatobecker 

All the above-mentioned tasks are completed. I think that the Livechat Widget part for this Feedback/Survey feature is complete. I'd appreciate if u could review this PR and any suggestions and feedback is most welcome:smile: 

**A bit about the feature implemented**

When the user clicks on `Finish this chat` option on the chat page, they will be redirected to the feedback page. 
Once the user gives the feedback and clicks `Send`, the app calls [`chatSurvey`](https://github.com/RocketChat/Rocket.Chat.js.SDK/blob/07a8bdc93c1e72410263a88b301d2bb6801b4ee5/src/lib/api/Livechat.ts#L31) method of RC SDK.
This is the response I get from the SDK
![image](https://user-images.githubusercontent.com/34130764/83157384-28e04300-a121-11ea-910d-c5ab7d54fb1f.png)
( I don't know how to verify whether is feedback has been saved in the main RC server. However, the `success` property in the above response most probably indicates that the operation is a success )

After giving feedback, the user is redirected to `Chat-Finished` Page.

renatobecker commented:
> Hello, @renatobecker
> 
> All the above-mentioned tasks are completed. I think that the Livechat Widget part for this Feedback/Survey feature is complete. I'd appreciate if u could review this PR and any suggestions and feedback is most welcome😄
> 
> **A bit about the feature implemented**
> 
> When the user clicks on `Finish this chat` option on the chat page, they will be redirected to the feedback page.
> Once the user gives the feedback and clicks `Send`, the app calls [`chatSurvey`](https://github.com/RocketChat/Rocket.Chat.js.SDK/blob/07a8bdc93c1e72410263a88b301d2bb6801b4ee5/src/lib/api/Livechat.ts#L31) method of RC SDK.
> This is the response I get from the SDK
> ![image](https://user-images.githubusercontent.com/34130764/83157384-28e04300-a121-11ea-910d-c5ab7d54fb1f.png)
> ( I don't know how to verify whether is feedback has been saved in the main RC server. However, the `success` property in the above response most probably indicates that the operation is a success )
> 
> After giving feedback, the user is redirected to `Chat-Finished` Page.

@murtaza98 I didn't review your PR in-depth yet, but I noticed the following situations:
- The colors/theme aren't being applied to the `Survey Feedback` Form/Container;
- When clicking on `No, thanks` button, nothing happens;
- Why did you use the `emoji-mart` lib if the original PR doesn't?

I'll be waiting for your reply, thanks.

renatobecker commented on src/components/Form/ExperienceRating/index.js:
```diff
@@ -0,0 +1,67 @@
+					createClassName(styles, 'form__input-experience-rating'),
+				].join(' ')}
+			>
+				<ExperienceRatingItem value="1" label={this.renderEmoji('white_frowning_face')} description="Very bad" onChange={this.handleChange} checked={this.isChecked('1')} />
```
You can use an Array/Map function to render `ExperienceRatingItem` from an array of objects that contains the rating items.

murtaza98 replied:
Made modifications to use array maps to render the `ExperienceRatingItem` component

renatobecker commented on src/components/Form/ExperienceRating/style.scss:
```diff
@@ -0,0 +1,294 @@
+@import '~styles/colors';
```
Do you really need all of this CSS / Styling stuff? 
I took a look at the original PR and `style.scss` file there is much more simple.

murtaza98 replied:
U r right. We don't need this. Removed it in the latest commit

renatobecker commented on src/lib/room.js:
```diff
@@ -14,6 +14,10 @@ const commands = new Commands();
 
 export const closeChat = async () => {
 	await handleTranscript();
+	route('/survey-feedback');
```
I don't like to code that. I prefer to implement a new configuration on the backend side that defines whether the `Survey Feedback` feature is enabled or not.

Let's think about it.

murtaza98 replied:
Hi @renatobecker,
I like this idea. This would be like the `Transcript` feature which Livechat already has. I think the admin should have a setting at the main server, to On/Off the Feedback section. This would be more powerful.

renatobecker replied:
> Hi @renatobecker,
> I like this idea. This would be like the `Transcript Chats` feature which Livechat already has. I think the admin should have a setting at the main server, to On/Off the Feedback section. This would be more powerful.

Yeah, that's the idea, we gonna provide a setting to enable/disable the survey feedback.

murtaza98 replied:
Hello, 
Added a PR in the main server which will add a setting to show/hide a Feedback form. Also updated the code here to connect with the backend setting.
Link to PR - https://github.com/RocketChat/Rocket.Chat/pull/17838

renatobecker commented on src/routes/SurveyFeedback/component.js:
```diff
@@ -0,0 +1,100 @@
+		message: [],
+	}
+
+	getValidableFields = () => Object.keys(this.validations)
```
The validation process has been changed, please take a look at the changes I did here in order to use the validations properly:

https://github.com/RocketChat/Rocket.Chat.Livechat/pull/419/files#diff-40c607440b11d922d97e83430f54daa8R24

murtaza98 replied:
Incorporated the validation process from the suggested PR here

renatobecker commented on src/routes/SurveyFeedback/container.js:
```diff
@@ -0,0 +1,80 @@
+				}
+				// console.log('Packet --> ', packet);
+				Livechat.chatSurvey(packet).then(response => {
+					console.log('SDK response - survey --> ', response);
```
Please, remove the console.log here.

murtaza98 replied:
Removed the console.log

renatobecker commented on yarn.lock:
```diff
@@ -5348,6 +5348,14 @@ elliptic@^6.0.0:
     minimalistic-assert "^1.0.0"
     minimalistic-crypto-utils "^1.0.0"
 
+emoji-mart@^3.0.0:
```
I don't think we need it.

murtaza98 replied:
removed the emoji-mart dependency :v:

murtaza98 commented:
Hi @renatobecker 
> * Why did you use the `emoji-mart` lib if the original PR doesn't?

When I tested the original PR, it emojis which were getting rendered were colorless. Have a look
![image](https://user-images.githubusercontent.com/34130764/83178555-4a9bf300-a13e-11ea-9b19-7bc25e2bbb23.png)

> * The colors/theme aren't being applied to the `Survey Feedback` Form/Container;
> * When clicking on `No, thanks` button, nothing happens;

I will work on these 2, thanks for pointing it out

renatobecker commented:
> Hi @renatobecker
> 
> > * Why did you use the `emoji-mart` lib if the original PR doesn't?
> 
> When I tested the original PR, the emojis which were getting rendered were colorless. Have a look
> ![image](https://user-images.githubusercontent.com/34130764/83178555-4a9bf300-a13e-11ea-9b19-7bc25e2bbb23.png)
> The SS by the owner of original PR #104, shows colorful emoji's rendered. I couldn't figure out y they were appearing colorless on my laptop, that's y I added `emoji-mart` library.
> 
> > * The colors/theme aren't being applied to the `Survey Feedback` Form/Container;
> > * When clicking on `No, thanks` button, nothing happens;
> 
> I will work on these 2, thanks for pointing it out

Another important thing to mention: You hardcoded the  `satisfaction` survey option, but you need to iterate through all survey options available on the `config` options.

murtaza98 commented:
> * The colors/theme aren't being applied to the `Survey Feedback` Form/Container;

Applied the color scheme to the feedback page

> * When clicking on `No, thanks` button, nothing happens;

Added a new behavior, now when `No, thanks` button is clicked, the user is directed to `chat-finished` route, and no feedback is recorded

> * Why did you use the `emoji-mart` lib if the original PR doesn't?

Removed the dependency of emoji-mart lib. The loading of emojis is now much faster.

> * Another important thing to mention: You hardcoded the satisfaction survey option, but you need to iterate through all survey options available on the config options.

Implemented this, now all the options are being read from the config. Nothing is hardcoded.

PS.
Thanks for the code review, I learned a lot from it :smile:

antkaz commented on src/routes/SurveyFeedback/container.js:
```diff
@@ -0,0 +1,97 @@
+					rid: rid,
+					data: data
+				}
+				Livechat.chatSurvey(packet).then(response => {
```
Hello! I checked this PR and found an error that the comment message is not sent to the server. It depends on the server config for livechat.

murtaza98 replied:
Hello @antkaz, U r correct. This PR doesn't send the comment message to the server. I tried sending it, but the server is not configured to handle it hence it was throwing an error. So in order to send the comment message, some of the interfaces and checks on the server will need to be modified.

antkaz replied:
I can help you do it

renatobecker replied:
It would be awesome!

antkaz replied:
I created new PR for ROcket.Chat https://github.com/RocketChat/Rocket.Chat/pull/18280 where I fixed old survey config and added feedback form config.
Also I created PR https://github.com/murtaza98/Rocket.Chat.Livechat/pull/2 where I fixed current code and added sending comment to Rocket.Chat server.
Feedback saves to room data 
![image](https://user-images.githubusercontent.com/19654261/87652897-394e8a80-c76e-11ea-8e6a-76ff193caa49.png)

antkaz replied:
@renatobecker @murtaza98 can you review please
